daphne/src/testing.rs

```diff
@@ -470,32 +474,156 @@ impl DapLeader<BearerToken> for MockAggregator {
+        let task_config = self
+            .get_task_config_for(&collect_req.task_id)
+            .ok_or_else(|| DapError::fatal("no task found"))?;
+
+        let mut leader_state_store_mutex_guard = self
+            .leader_state_store
+            .lock()
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+        let leader_state_store = leader_state_store_mutex_guard.deref_mut();
+
+        // Store Collect ID and CollectReq into LeaderState.
+        let collect_id = Id(rng.gen());
+        if !leader_state_store.contains_key(&collect_req.task_id) {
+            leader_state_store.insert(collect_req.task_id.clone(), LeaderState::new());
+        }
+
+        let leader_state = leader_state_store.get_mut(&collect_req.task_id).unwrap();
+        leader_state.collect_ids.push_back(collect_id.clone());
+        let collect_job_state = CollectJobState {
+            collect_req: Some(collect_req.clone()),
+            collect_resp: None,
+        };
+        leader_state
+            .collect_jobs
+            .insert(collect_id.clone(), collect_job_state);
+
+        // Construct a new Collect URI for this CollectReq.
+        let collect_uri = task_config
+            .leader_url
+            .join(&format!(
+                "/collect/task/{}/req/{}",
+                collect_req.task_id.to_base64url(),
+                collect_id.to_base64url(),
+            ))
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+
+        Ok(collect_uri)
     }
 
+    // Called to retrieve completed CollectResp at the request of Collector.
     async fn poll_collect_job(
         &self,
-        _task_id: &Id,
-        _collect_id: &Id,
+        task_id: &Id,
+        collect_id: &Id,
     ) -> Result<DapCollectJob, DapError> {
-        unreachable!("not implemented");
+        let mut leader_state_store_mutex_guard = self
+            .leader_state_store
+            .lock()
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+        let leader_state_store = leader_state_store_mutex_guard.deref_mut();
+
+        let leader_state = leader_state_store
+            .get(task_id)
+            .ok_or_else(|| DapError::fatal("cannot get collect job for task_id"))?;
+        if leader_state.collect_jobs.contains_key(collect_id) {
+            if let Some(collect_resp) = &leader_state
+                .collect_jobs
+                .get(collect_id)
+                .unwrap()
+                .collect_resp
+            {
+                Ok(DapCollectJob::Done(collect_resp.clone()))
+            } else if leader_state
+                .collect_jobs
+                .get(collect_id)
+                .unwrap()
+                .collect_resp
+                .is_none()
+            {
+                Ok(DapCollectJob::Pending)
+            } else {
+                Ok(DapCollectJob::Unknown)
+            }
+        } else {
+            Err(DapError::fatal("cannot get collect job for collect_id"))
+        }
```

Comment:
Hmm, this doesn't look quite right. Is the outer `else` statement reachable? What about the inner `else`?

Taking a step back: Your goal here is to check if collect request is "finished", "pending", or "unknown". It might be helpful to try to first describe this algorithm to yourself in natural language, then try to transcribe that algorithm into Rust.

Reply:
Thanks for the suggestion and based on your feedback during our call, I've fixed this.

I've also added a test case that tests all these return types.